hudi-common/src/main/java/org/apache/hudi/common/util/queue/HoodieExecutor.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hudi.common.util.queue;
+
+import java.util.List;
+import java.util.concurrent.Future;
+
+/**
+ * HoodieExecutor which orchestrates concurrent producers and consumers communicating through a bounded in message queue.
+ */
+public interface HoodieExecutor<I, O, E> {
```

Comment:
I'd suggest to reserve this interface for APIs only exposed to the actual user: in that case such only APIs are `execute` and `close`. All the other APIs are implementation detils (setup, startConumers, postAction, etc) and shouldn't be exposed outside actual implementations: 
  - Let's move all of these to a shared base class
  - Let's keep them "protected"

Let's also make this extend `AutoCloseable`

Reply:
Nice catch!
Just keep
```
E execute();
boolean isRemaining();
void shutdownNow();
boolean awaitTermination();
```
in `public interface HoodieExecutor<I, O, E> extends Closeable` which are used to control the lifecycle for hoodieExecutor for example `SparkLazyInsertIterable#computeNext`

And move all other function into `HoodieExecutorBase` and changed to protect except `startProducers` and `getQueue` which need to be call outside.